### PR TITLE
[#17] Add lerna to support multiple packages

### LINKED ISSUE #17: Add lerna to support multiple packages
## Why

- This repository contains the base ESLint configuration, but we are going to add the React ESLint, then it should support having multiple packages in this repository.
- Use [lerna](https://lerna.js.org/) to manage them.

## Who Benefits?

Developers who use this shared ESLint configuration

### PR DESCRIPTION
**Pull request by hoangmirs**
Close https://github.com/nimblehq/eslint-config-nimble/issues/17

## What happened 👀

- Because we might have more shared ESLint configurations(base, react,...) on this repository, then I add [lerna](https://lerna.js.org/) to support multiple packages.
- Create a new README in the root folder to describe and add contribution guidelines.
- Copy the old content into `packages/eslint-config-nimble` and it will be the base configuration. Others will be based on it.
- Support [`workspaces`](https://docs.npmjs.com/cli/v8/using-npm/workspaces) to work with multiple packages.

## Insight 📝

- To publish a new version of a package, just need to run `npx lerna publish`. Lerna will detect the changes and list the package along with the version that needs to be published:
npx_lerna_publish
_Note that you need to log into the npm account on your local first._
- The new package should have the format: `@nimblehq/eslint-config-nimble-*` e.g. `@nimblehq/eslint-config-nimble-react`

## Proof Of Work 📹

The publish command should work, this is the tested tag that was published by `lerna`:
https://github.com/nimblehq/eslint-config-nimble/releases/tag/%40nimblehq%2Feslint-config-nimble%402.2.2-alpha.0

### CODE REVIEW

File: packages/eslint-config-nimble/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    // Additional, per-project rules...
+  }
+}
+```
+
+This would require defining the required dependencies in the project itself.
+
+## License
+
+This project is Copyright (c) 2014 and onwards.
```
Comment (malparty) [inline packages/eslint-config-nimble/README.md]:
```suggestion
This project is Copyright (c) 2014 and onwards Nimble.
```
I think a `Copyright` mention always comes with the author's name, isn't it? 🤔

File: README.md [hunk 2]
```diff
@@ -2,64 +2,54 @@
...
 
-### With a framework
+### Add a new package
+
+- Use [`lerna create`](https://github.com/lerna/lerna/tree/main/commands/create#readme) command to add a new package.
+
+- The package name should have the following format: `@nimblehq/{package-name}`. This project is shareable configurations for [ESLint](https://eslint.org/), the `{package-name}` should be `eslint-config-nimble-*`:
+
+```bash
+  npx lerna create @nimblehq/eslint-config-nimble-react
```
Comment (malparty) [inline README.md]:
```suggestion
  lerna create @nimblehq/eslint-config-nimble-react
```

If we recommend to install `lerna`, no need to use npx. Or earlier in this README, we can remov the "Install Lerna for accessing to the lerna CLI" ;-)

**Review by malparty — APPROVED** [review]
2 minor comments, very interesting evolution for this package!! :) Thanks!

File: packages/eslint-config-nimble/README.md (hunk 1 above)
Comment (hoangmirs) [reply packages/eslint-config-nimble/README.md]: I am also not sure about that 😂 
@nimblehq/engineering-leads Could you give us a suggestion? Other repositories are using this text as well 🙏

File: README.md (hunk 2 above)
Comment (hoangmirs) [reply README.md]: I removed the npx instead 😂

File: packages/eslint-config-nimble/README.md (hunk 1 above)
Comment (olivierobert) [reply packages/eslint-config-nimble/README.md]: @malparty is correct. It should also be fixed on https://github.com/nimblehq/git-template

**Comment by olivierobert** [comment]
@malparty @hoangmirs  FYI on [the discussion regarding NPM and Yarn](https://github.com/nimblehq/react-templates/issues/66), the latter has a feature called Workspaces that does the same as Lerna. Here is [a tutorial ](https://semaphoreci.com/blog/javascript-monorepos-yarn-workspaces) and[ the official docs](https://yarnpkg.com/cli/workspace).

**Comment by hoangmirs** [comment]
> @malparty @hoangmirs FYI on [the discussion regarding NPM and Yarn](https://github.com/nimblehq/react-templates/issues/66), the latter has a feature called Workspaces that does the same as Lerna. Here is [a tutorial ](https://semaphoreci.com/blog/javascript-monorepos-yarn-workspaces) and[ the official docs](https://yarnpkg.com/cli/workspace).

Thank you for your information. And the good news is the recent npm version also supports [workspaces feature](https://docs.npmjs.com/cli/v7/using-npm/workspaces).
But they are different and we can use both of them (lerna and workspaces) in our project:
- Workspaces: Work with multiple packages easier, it also uses the root `node_modules` for all packages to reduce the installation time.
- Lerna: The most important command of Lerna is `lerna publish` - publish will help publish any updated packages.

I will add workspace to this project 👍

File: packages/eslint-config-nimble/README.md (hunk 1 above)
Comment (malparty) [reply packages/eslint-config-nimble/README.md]: Fixed in https://github.com/nimblehq/git-template/pull/24 (I did not create an issue as this is quite minor 🤔 )

Comment (hoangmirs) [reply packages/eslint-config-nimble/README.md]: I also fixed it in https://github.com/nimblehq/eslint-config-nimble/pull/20/commits/4bbfe48a7e89269a56cd3349c6d8b261fda38450

**Review by olivierobert — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -2,64 +2,69 @@
...
 > ESLint [shareable config](http://eslint.org/docs/developer-guide/shareable-configs.html) developed and maintained by [nimble](https://nimblehq.co/)
 
-## Installation
+## Usage
 
-```
-$ npm install --save-dev @nimblehq/eslint-config-nimble
-```
+The configurations are separated into dedicated packages:
+- [eslint-config-nimble](/packages/eslint-config-nimble): ESLint base rules
```
Comment (carryall) [inline README.md]:
```suggestion
The configurations are separated into dedicated packages:

- [eslint-config-nimble](/packages/eslint-config-nimble): ESLint base rules
```

File: README.md [hunk 4]
```diff
@@ -2,64 +2,69 @@
...
+```
+
+_The current branch that you run the publish command should be pushed on Github._
+
+### Run commands
+
+This project also uses [workspaces](https://docs.npmjs.com/cli/v8/using-npm/workspaces) feature.
+
+To run a command:
+- Add the command to `scripts` part in `package.json` files.
```
Comment (carryall) [inline README.md]:
```suggestion
To run a command:

- Add the command to `scripts` part in `package.json` files.
```

File: packages/eslint-config-nimble/README.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
+# @nimblehq/eslint-config-nimble
+
+This package provides ESLint base rules
+
+## Installation
+
+```
+$ npm install --save-dev @nimblehq/eslint-config-nimble
+```
```
Comment (carryall) [inline packages/eslint-config-nimble/README.md]:
```suggestion
```bash
  npm install --save-dev @nimblehq/eslint-config-nimble
```
```

File: packages/eslint-config-nimble/README.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    // Additional, per-project rules...
+  }
+}
+```
+
+### With a framework
+
+Similar to the process above, but usually it requires adding the extra rules for the JS framework:
+
+
```
Comment (carryall) [inline packages/eslint-config-nimble/README.md]:
```suggestion

```

**Review by carryall — APPROVED** [review]
Only some small comments with space adjustment 😅

File: README.md (hunk 3 above)
Comment (hoangmirs) [reply README.md]: Fix this and other comments in https://github.com/nimblehq/eslint-config-nimble/pull/20/commits/d74720d74295f92913a3eb8f5efc67fcb5b07783

**Pull request merged**